System/Metrics/Distribution.hs

```diff
@@ -0,0 +1,442 @@
+import System.Metrics.ThreadId
+
+#if WORD_SIZE_IN_BITS >= 64
+
+import Data.Primitive.ByteArray
+import GHC.Float
+import GHC.Int
+import GHC.IO
+import GHC.Prim
+import Test.Inspection
+
+#else
+
+import qualified Prelude
+
+import Control.Monad ((<$!>), when)
+import Data.Atomics (atomicModifyIORefCAS_)
+import Data.IORef
+import Data.Int (Int64)
+
+#endif
+
+------------------------------------------------------------------------
+
+-- | An metric for tracking events.
+newtype Distribution = Distribution { unD :: A.Array Stripe }
+
+-- | Number of lock stripes. Should be greater or equal to the number
+-- of HECs.
+numStripes :: Int
+numStripes = 8
+
+-- | Get the stripe to use for this thread.
+myStripe :: Distribution -> IO Stripe
+myStripe distrib = do
+    tid <- myCapability
+    return $! unD distrib `A.index` (tid `mod` numStripes)
+
+------------------------------------------------------------------------
+
+-- We want at least 64 bits in order to avoid overflow of the count of
+-- samples added to the distribution.
+
+#if WORD_SIZE_IN_BITS >= 64
+
+-- If the machine word size is 64 bits, we use a spinlock and ensure
+-- that the code that runs while holding the lock cannot be preempted by
+-- the runtime system. This is achieved by using only (non-allocating)
+-- GHC prim ops when holding the lock.
+--
+-- We pad to 64-bytes (an x86 cache line) to try to avoid false sharing.
+
+newtype Stripe = Stripe (MutableByteArray RealWorld)
+
+sIZEOF_CACHELINE :: Int
+sIZEOF_CACHELINE = 64
+{-# INLINE sIZEOF_CACHELINE #-}
+
+posLock, posCount, posMean, posSumSqDelta, posSum, posMin, posMax :: Int
+-- We put the variable-sized `Int` field first so that its offset (0)
+-- does not depend on its size.
+posLock = 0 -- Int
+posCount = 1 -- Int64
+posMean = 2 -- Double
+posSumSqDelta = 3 -- Double
+posSum = 4 -- Double
+posMin = 5 -- Double
+posMax = 6 -- Double
+{-# INLINE posLock #-}
+{-# INLINE posCount #-}
+{-# INLINE posMean #-}
+{-# INLINE posSumSqDelta #-}
+{-# INLINE posSum #-}
+{-# INLINE posMin #-}
+{-# INLINE posMax #-}
+
+newStripe :: IO Stripe
+newStripe = do
+    -- We pad to 64 bytes (an x86 cache line) to try to avoid false sharing.
+    arr <- newAlignedPinnedByteArray sIZEOF_CACHELINE sIZEOF_CACHELINE
+
+    writeByteArray @Int    arr posLock        0
+    writeByteArray @Int64  arr posCount       0
+    writeByteArray @Double arr posMean        0.0
+    writeByteArray @Double arr posSumSqDelta  0.0
+    writeByteArray @Double arr posSum         0.0
+    writeByteArray @Double arr posMin         inf
+    writeByteArray @Double arr posMax         (-inf)
+
+    pure $ Stripe arr
+  where
+    inf :: Double
+    inf = 1/0
+
+spinlock#
+  :: MutableByteArray# RealWorld -> State# RealWorld -> State# RealWorld
+spinlock# arr s0 =
+  case posLock of { I# posLock' ->
+  case casIntArray# arr posLock' 0# 1# s0 of { (# s1, r #) ->
+  case r of
+    0# -> s1
+    _ -> spinlock# arr s1
+  }}
```

Comment:
I'm always a little wary whenever I see spinlocks, but I'm not 100% sure about the constraints you're working with here so in lieu of offering targeted advice I'll defer to [this blog post by Aleksey Kladov](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html) as well as his [follow-up indicating that "sleeping mutexes" should be faster anyway](https://matklad.github.io/2020/01/04/mutexes-are-faster-than-spinlocks.html).

Reply:
Thanks for showing me the dangers of spinlocks! To be honest, I don't know what I'm doing when it comes to spinlocks or locking in general. This code is a translation of and replacement for some C code that was previously in this `ekg-core` library (and the replacement was motivated by reports that the C code was generating nonsensical results on aarch64).

Reply:
I would like to postpone replacing the spinlock, or any other work on the `Distribution` metric type. Rather, I think we should replace it entirely, maybe with Prometheus' histograms and summaries?